src/core/symbology/qgsfillsymbol.cpp

```diff
@@ -41,7 +41,7 @@ void QgsFillSymbol::renderPolygon( const QPolygonF &points, const QVector<QPolyg
   const double opacity = dataDefinedProperties().hasActiveProperties() ? dataDefinedProperties().valueAsDouble( QgsSymbol::Property::Opacity, context.expressionContext(), mOpacity * 100 ) * 0.01
                          : mOpacity;
 
-  QgsSymbolRenderContext symbolContext( context, Qgis::RenderUnit::Unknown, opacity, selected, mRenderHints, f );
+  QgsSymbolRenderContext symbolContext( context, Qgis::RenderUnit::Unknown, opacity >= 0 && opacity <= 1 ? opacity : mOpacity, selected, mRenderHints, f );
```

Comment:
Actually, I wonder if we shouldn't just be clamping the `opacity` value here to 0-1. That way if a data defined property is set where the expression results in a value like 100.0000001 we won't just fallback to the default opacity...

Reply:
My gut feeling is that clamping the data defined value is the correct approach, but I'm not 100% on that. @nirvn got any thoughts here?

Reply:
@nyalldawson looking better at the current behavior it looks like I was wrong... you are right! You proposed approach to clamp opacity to [0-1] here is better and gives a rendering more similar to the current one, although not the very same.
Since it was not my initial intention to change this behavior, but only to correct the display of the symbology preview, I'me more inclined to remove the second commit.
Let me know what do you thing.

Reply:
@agiudiceandrea I think we should handle this gracefully, so even in debug builds we don't get the warnings :) 

So I'd be in favour of calling std::clamp on the opacity prior to using it.

Reply:
@nyalldawson , sounds good to me.

Reply:
I've removed the second added commit. I think it would be possible to merge this PR with the remaining only commit which fixes an actual bug.
I prefer to propose another PR for master only to add the clamping for every setAlphaF in Qgs*Symbol to avoid the Qt warning messages, although I have a concern that this may negatively affect the rendering times just to avoid the warning messages.